src/entropy.rs

```diff
@@ -0,0 +1,407 @@
+            .apply(|result, &p, &q| {
+                *result = {
+                    if p == A::zero() {
+                        A::zero()
+                    } else {
+                        p * (q / p).ln()
+                    }
+                }
+            });
+        let kl_divergence = temp.sum();
+        Ok(Some(-kl_divergence))
+    }
+
+    fn cross_entropy<S2>(&self, q: &ArrayBase<S2, D>) -> Result<Option<A>, ShapeMismatch>
+    where
+        S2: Data<Elem=A>,
+        A: Float,
+    {
+        if self.len() == 0 {
+            return Ok(None)
+        }
+        if self.shape() != q.shape() {
+            return Err(ShapeMismatch {
+                first_shape: self.shape().to_vec(),
+                second_shape: q.shape().to_vec()
+            })
+        }
+
+        let mut temp = Array::zeros(self.raw_dim());
+        Zip::from(&mut temp)
+            .and(self)
+            .and(q)
+            .apply(|result, &p, &q| {
+                *result = {
+                    if p == A::zero() {
+                        A::zero()
+                    } else {
+                        p * q.ln()
+                    }
+                }
+            });
+        let cross_entropy = temp.sum();
+        Ok(Some(-cross_entropy))
+    }
+}
+
+#[cfg(test)]
+mod tests {
+    use super::EntropyExt;
+    use std::f64;
+    use approx::assert_abs_diff_eq;
+    use noisy_float::types::n64;
+    use ndarray::{array, Array1};
+    use errors::ShapeMismatch;
+
+    #[test]
+    fn test_entropy_with_nan_values() {
+        let a = array![f64::NAN, 1.];
+        assert!(a.entropy().unwrap().is_nan());
+    }
+
+    #[test]
+    fn test_entropy_with_empty_array_of_floats() {
+        let a: Array1<f64> = array![];
+        assert!(a.entropy().is_none());
+    }
+
+    #[test]
+    fn test_entropy_with_array_of_floats() {
+        // Array of probability values - normalized and positive.
+        let a: Array1<f64> = array![
+            0.03602474, 0.01900344, 0.03510129, 0.03414964, 0.00525311,
+            0.03368976, 0.00065396, 0.02906146, 0.00063687, 0.01597306,
+            0.00787625, 0.00208243, 0.01450896, 0.01803418, 0.02055336,
+            0.03029759, 0.03323628, 0.01218822, 0.0001873 , 0.01734179,
+            0.03521668, 0.02564429, 0.02421992, 0.03540229, 0.03497635,
+            0.03582331, 0.026558  , 0.02460495, 0.02437716, 0.01212838,
+            0.00058464, 0.00335236, 0.02146745, 0.00930306, 0.01821588,
+            0.02381928, 0.02055073, 0.01483779, 0.02284741, 0.02251385,
+            0.00976694, 0.02864634, 0.00802828, 0.03464088, 0.03557152,
+            0.01398894, 0.01831756, 0.0227171 , 0.00736204, 0.01866295,
+        ];
+        // Computed using scipy.stats.entropy
+        let expected_entropy = 3.721606155686918;
+
+        assert_abs_diff_eq!(a.entropy().unwrap(), expected_entropy, epsilon = 1e-6);
+    }
+
+    #[test]
+    fn test_cross_entropy_and_kl_with_nan_values() -> Result<(), ShapeMismatch> {
```

Comment:
Woah, I didn't know that we can return `Result` from tests now. Cool!

Reply:
I thought it was just for `main`, but it turns out it works for tests as well :sunglasses: